base/promotion.jl

```diff
@@ -33,7 +33,7 @@ function typejoin(@nospecialize(a), @nospecialize(b))
         if !(b <: Tuple)
             return Any
         end
-        ap, bp = a.parameters, b.parameters
+        ap, bp = a.parameters::Core.SimpleVector, b.parameters::Core.SimpleVector
```

Comment:
Since we implicitly assume here that `a::DataType`, perhaps we should move the type-assert code above this `elseif a <: Tuple` branch. And turn this into a `if a <: Tuple` there.
```
    # We have to hide Constant info from inference, see #44390
     a, b = inferencebarrier(a)::DataType, inferencebarrier(b)::DataType
```